Add alpha 3 changelog. (#1411)
edgedb/edgedb: docs/changelog/1_0_a3.rst

===========
1.0 Alpha 3
===========

This changelog summarizes new features and breaking changes in
EdgeDB 1.0 alpha 3.

CLI
===

> We've converted EdgeDB command-line tools to using Rust. The tools are

Reviewer: ```suggestion
The EdgeDB command-line tools are now written in Rust. The tools are
```

Author: OK

a single binary now and can run on multiple different platforms,
including Windows.

Notably the REPL has been converted and now features more commands:

.. code-block::

      \dump FILENAME           dump current database into a file
      \restore FILENAME        restore the database from file into
                               the current one

>       \create-database DBNAME  create a new database

Reviewer: Please drop `\create-database`

Author: Done

                               in default output mode (may fail if
                               schema is updated after enabling
                               option)
    limit                      Set implicit LIMIT. Defaults to 100,
                               specify 0 to disable
    output-mode                Set output mode. One of: json, json-
                               elements, default, tab-separated
    verbose-errors             Print all errors with maximum
                               verbosity

EdgeQL
======

* Add ``+=`` and ``-=`` operations for use in :eql:stmt:`UPDATE`
  (:eql:gh:`#165`).
* Distinguish ``REQUIRED`` and ``OPTIONAL`` query parameters
  (:eql:gh:`#1352`, :eql:gh:`#1355`)
* Allow tuple arrays in schema definitions (:eql:gh:`73125882`).
* Get rid of ``schema::CompoundType`` in favor of ``is_compound_type``
  property (:eql:gh:`c30c1c4f`).
* Add :eql:type:`anytype` and :eql:type:`anytuple` as proper abstract
  types to the schema (:eql:gh:`affb65de`).
* ``schema::Constraint`` no longer has ``args``, use ``params`` instead
  (:eql:gh:`14cec6f1`).
* Make ``std::Object`` the implicit base type for all user-defined

>   object types only.

Reviewer: ```suggestion
  object types.
```

Author: I put "only" there to emphasize how this is different from the `BaseObject` that's listed onthe next line.

* Accept ``1e100n`` format as a valid ``bigint`` constant rather than
  ``decimal`` (:eql:gh:`#1372`)
* Make self-referencing (recursive) aliases forbidden.
* Fix dependency tracking affecting index creation (:eql:gh:`#1181`).
* Fix derivation of link targets in certain cases of multiple
  inheritance (:eql:gh:`52c6b2d4`).
* Fix handling of ad-hoc tuples (:eql:gh:`#1255`).
* Fix incorrect implicit limit injection in subqueries in computables
  (:eql:gh:`#1271`).
* Computables cardinality must now be declared explicitly as
  ``required``, ``single`` or ``multi``. The expression is validated
  to be within the upper and lower limits implied by the declaration
  (:eql:gh:`#1201`, :eql:gh:`#349`).
* Prohibit database names longer than 63 characters (:eql:gh:`#1158`).
* Prohibit duplicate elements in shapes (:eql:gh:`#1368`).

GraphQL
=======

* Reflect "description" annotations into GraphQL descriptions
  (:eql:gh:`#1228`).
* Change the auto-generated type name template to use "\_Type" suffix
  (:eql:gh:`#1175`).
* Fix incorrect reflection of enums (:eql:gh:`#1227`).
* Fix deeply nested insert issue (:eql:gh:`#1243`).

Standard Library
================

> * Add ``sys::get_current_database()``

Reviewer: Can you use `eql:func:` here?

Author: Done

* Rename ``std::to_str()`` array-joining function to
  :eql:func:`array_join`.
* Rename ``std::to_array()`` string-splitting function to
  :eql:func:`str_split`.

Server
======

* Convert EdgeQL lexer (:eql:gh:`#1178`)
* Convert GraphQL parser to Rust and perform constant extraction
  (:eql:gh:`#1299`).
* Constant extraction from EdgeQL queries (:eql:gh:`#1356`)
* Make it possible to use different EdgeDB servers connected to the
  same Postgres cluster (:eql:gh:`#1197`).
* Protocol: implement ``JSON_ELEMENTS`` IO format for responses
  (:eql:gh:`#1169`).

Misc
====

> * Add cheatsheet to documentation.

Reviewer: I'd hyperlink the "cheatsheet"

Author: OK, done.

* Strictly type-annotate the core schema module.
* Clean-up and improve schema and introspection code to make the
  internal APIs more consistent (:eql:gh:`#1408`).
* Update the `edgedb-js <https://github.com/edgedb/edgedb-js>`_ driver
  to v0.7.3.
* Update the `edgedb-python <https://github.com/edgedb/edgedb-python>`_

>   driver to v0.8.0a1.

Reviewer: 0.8.0, without a1.  We'll release a stable version today/tomorrow.

Author: OK

===========
1.0 Alpha 3
===========

This changelog summarizes new features and breaking changes in
EdgeDB 1.0 alpha 3.

> Rust

Reviewer: I'd call this section CLI

Author: I thought that "Rust" might highlight the nature of the change better, but we can make it CLI instead (and move the comment about internals rewritten in Rust to "Server")

Reviewer: Well, we want the changelog to be succinct and to the point. The section is about CLI, and its rewrite to Rust is an important point, but not the only point.

====

We've started the process of rewriting some of the
performance-critical EdgeDB internals to Rust. So far we have
converted the lexer as well as the GraphQL parser.

> Additionally, we're also making EdgeDB command-line tools using Rust.

Reviewer: making...using --- this sentence needs a rewrite. I'd also mention the command line tools first and then talk about graphql.

Author: OK, I'll swap order.

> Notably the REPL has been converted and now features more commands:

Reviewer: All tools have been converted. The goal of the conversion is ease of distribution. The tools are a single binary now and can run on multiple different platforms, including Windows.

Author: I'll add the comment about distribution, etc.

Reviewer: "Notably the REPL has been converted and now features more commands:" -- this sentence is not needed anymore?

Victor, please do a complete read of this file and fix it.

Author: Reworded.

> .. code-block::

Reviewer: I don't think we need to include this print-out.  Besides, there are only a few new commands (dump/restore?)

Author: History, edit, set, recent error are also new. Plus the commands themselves changed (and we've had them listed last time).

Reviewer: I'd just enumerate the new commands then, briefly. Like "there are new commands to manage and navigate history: `\blah1`, `\blah2`"

      \lT [-sI] [PATTERN]      list scalar types
      \lt [-sI] [PATTERN]      list object types
      \lr [-I]                 list roles
      \lm [-I]                 list modules
      \la [-Isv] [PATTERN]     list expression aliases
      \lc [-I] [PATTERN]       list casts
      \li [-Isv] [PATTERN]     list indexes
      \list-ports              list ports

    Operations
      \dump FILENAME           dump current database into a file
      \restore FILENAME        restore the database from file into
                               the current one
      \create-database DBNAME  create a new database

    Editing
      \s, \history             show history
      \e, \edit [N]            spawn $EDITOR to edit history entry N
                               then use the output as the input

    Settings
      \set [OPTION [VALUE]]    show/change setting

    Connection
      \c, \connect [DBNAME]    Connect to database DBNAME

    Development
      \E                       show most recent error message at
                               maximum verbosity
      \pgaddr                  show the network addr of the postgres
                               server

>       \psql                    open psql to the current postgres

Reviewer: This is also old output, we dropped this command from help text.

Author: Odd, I used my REPL help to get this (which should be pretty fresh). I'll pull the latest version.

  inheritance (:eql:gh:`52c6b2d4`).
* Fix handling of ad-hoc tuples (:eql:gh:`#1255`).
* Fix incorrect implicit limit injection in subqueries in computables
  (:eql:gh:`#1271`).
* Computables cardinality must now be declared explicitly as
  ``required``, ``single`` or ``multi``. The expression is validated
  to be within the upper and lower limits implied by the declaration
  (:eql:gh:`#1201`, :eql:gh:`#349`).
* Prohibit database names longer than 63 characters (:eql:gh:`#1158`).
* Prohibit duplicate elements in shapes (:eql:gh:`#1368`).

GraphQL
=======

* Reflect "description" annotations into GraphQL descriptions
  (:eql:gh:`#1228`).
* Change the auto-generated type name template to use "\_Type" suffix
  (:eql:gh:`#1175`).
* Fix incorrect reflection of enums (:eql:gh:`#1227`).
* Fix deeply nested insert issue (:eql:gh:`#1243`).

Standard Library
================

* Add ``sys::get_current_database()``
* Add ``std::to_array()`` string-splitting function.

Server
======

* Make it possible to use different EdgeDB servers connected to the
  same Postgres cluster.
* Protocol: implement ``JSON_ELEMENTS`` IO format for responses
  (:eql:gh:`#1169`).

Misc
====

> * Add cheatsheet.

Reviewer: ... to the documentation.

> * Add an interactive tutorial.

Reviewer: Not related to EdgeDB changelog.

* Strictly type-annotate the core schema module.

> * Clean-up and improve schema and introspection code.

Reviewer: Please reference Elvis' PR.

Author: There's more than one PR that cleans up schema code (like making collections more well-behaved, or generalizing introspection). Collectively they make schema and introspection work better, but I'm not sure that pointing to any individual PR is really all that enlightening.

Reviewer: Please reference this pr https://github.com/edgedb/edgedb/pull/1408

> I'm not sure that pointing to any individual PR is really all that enlightening.

I'm not then sure why we are referencing any commits/prs?

* A number of updates and bugfixes for Python and JavaScript drivers,

>   including the support of ``edgedb-js`` for browsers without BigInt.

Reviewer: I wouldn't mention browser support in `edgedb-js` -- it's not even alpha stage.

Reviewer: Also if there are some concrete issues/prs for drivers please reference them.
